### PR TITLE
[WIP] Multiple Scala versions for kernel

### PR DESCRIPTION
**Pull request by jeremyrsmith**
- Update the distribution to include kernel packages for every supported Scala version (currently 2.11 and 2.12).
- Add configuration for setting the Scala version
- Remove shapeless from polynote-kernel (turns out it is 5MB of dependencies!) and instead just reimplement the few pieces that were used

TODO:
- UI for setting the Scala version per-notebook?
- More fiddling with the distribution to make it smaller (e.g. maybe we don't need to ship the server app for two different Scala versions)
- Remove the `KernelIsolation.Never` case, because it currently does nothing
- Test the updated launch script in vivo

### CODE REVIEW

File: scripts/polynote.py [hunk 1]
```diff
@@ -7,6 +7,10 @@
 if not os.environ.get('PYTHONPATH'):
     os.environ['PYTHONPATH'] = sys.prefix
 
+scala_version = os.environ.get('POLYNOTE_SCALA_VERSION')
```
Comment (jonathanindig) [inline scripts/polynote.py:10]: You can provide a default value: `os.environ.get('POLYNOTE_SCALA_VERSION', '2.11')`

File: build.sbt [hunk 2]
```diff
@@ -132,6 +154,9 @@ val `polynote-kernel` = project.settings(
     "net.sf.py4j" % "py4j" % "0.10.7",
     "org.scalamock" %% "scalamock" % "4.4.0" % "test"
   ),
+  distFiles := Seq(assembly.value) ++ (dependencyClasspath in Compile).value.collect {
+    case jar if jar.data.name.matches(".*scala-(library|reflect|compiler|collection-compat|xml).*") => jar.data
+  },
```
Comment (jonathanindig) [inline build.sbt:159]: Is this part basically for packaging scala library jars?

File: polynote-kernel/src/main/scala/polynote/kernel/remote/DeploySparkSubmit.scala [hunk 3]
```diff
@@ -70,18 +71,19 @@ object DeploySparkSubmit extends DeployCommand {
...
   } yield build(
     config,
     nbConfig,
     path,
+    classPath.map(_.toUri.toURL),
     serverArgs =
       "--address" :: serverAddress.getAddress.getHostAddress ::
       "--port" :: serverAddress.getPort.toString ::
-      "--kernelFactory" :: classOf[LocalSparkKernelFactory].getName ::
+      "--kernelFactory" :: "polynote.kernel.LocalSparkKernelFactory" ::
```
Comment (jonathanindig) [inline polynote-kernel/src/main/scala/polynote/kernel/remote/DeploySparkSubmit.scala:86]: Why this change?

**Comment by jonathanindig** [comment]
Awesome! A few questions:

1. Did removing the shapeless stuff fix that static initializer issue (the "run it again and it'll work" one)? 
2. Should we add the Scala version to the notebook metadata? 
3. UI would be great but I'm fine waiting for a subsequent PR if you like.

File: build.sbt (hunk 2 above)
Comment (jeremyrsmith) [reply build.sbt:159]: Yeah, it's saying what things we need to copy to the distribution from that module.

File: polynote-kernel/src/main/scala/polynote/kernel/remote/DeploySparkSubmit.scala (hunk 3 above)
Comment (jeremyrsmith) [reply polynote-kernel/src/main/scala/polynote/kernel/remote/DeploySparkSubmit.scala:86]: I moved `DeploySparkSubmit` to `polynote-kernel`, so it can't see the `LocalSparkKernelFactory` class anymore. I did that basically so that `Main` doesn't have to be in `polynote-spark` anymore.

**Comment by jeremyrsmith** [comment]
@jonathanindig 

> Did removing the shapeless stuff fix that static initializer issue (the "run it again and it'll work" one)?
I'm not sure, haven't gotten as far as testing yet (and that one has always been hard for me to reproduce)

> Should we add the Scala version to the notebook metadata?
There is already a notebook configuration item for specifying it at the notebook level, if that's what you mean?

> UI would be great but I'm fine waiting for a subsequent PR if you like.
Yeah, I think that will be in a separate PR once the UI is stable.

**Comment by jonathanindig** [comment]
@jeremyrsmith 

> > Did removing the shapeless stuff fix that static initializer issue (the "run it again and it'll work" one)?
> 
> I'm not sure, haven't gotten as far as testing yet (and that one has always been hard for me to reproduce)
 
I was able to reproduce as follows: in a fresh kernel, run `(0 to 10).map(i => (i, i+1)).toList`. The first time it runs, you will only get a `StringRepr` result. The second time you run, you will get a proper `StreamingRepr`. Note that this must be the first code the kernel runs, so a Spark kernel won't do this because it has a predef.

> > Should we add the Scala version to the notebook metadata?
> 
> There is already a notebook configuration item for specifying it at the notebook level, if that's what you mean?

Oops, missed that! 😅

**Review by jonathanindig — APPROVED** [review]
LGTM, though we should note that there are some backwards-incompatible changes here (kernel isolation `never` is gone (so some configs might start failing), some changes might be needed for development (no more `LocalSparkKernel`...). Perhaps a CHANGELOG entry is warranted?

File: polynote-kernel/src/main/scala/polynote/kernel/logging/package.scala [hunk 4]
```diff
@@ -142,6 +146,14 @@ package object logging {
           lastRemote.lazySet(null)
           printWithPrefix(warnPrefix, warnIndent, msg)
         }
+
+        override def warn(msg: Option[String], err: Throwable)(implicit location: Location): UIO[Unit] = blocking.effectBlocking {
+          out.synchronized {
+            lastRemote.lazySet(null)
+            logStackTraceSync(msg, err, warnPrefix, warnIndent)
```
Comment (jonathanindig) [inline polynote-kernel/src/main/scala/polynote/kernel/logging/package.scala:153]: How about making the color `Yellow`?

File: scripts/polynote.py [hunk 5]
```diff
@@ -18,20 +20,20 @@
...
 
-deps_path = Path(polynote_dir).joinpath("deps")
+deps_path = Path(polynote_dir).joinpath("deps", scala_version)
 
 if not(deps_path.exists()):
     raise Exception("Couldn't find the deps directory. Are we in the polynote installation directory?")
 
-deps = Path(polynote_dir).joinpath("deps").glob("*.jar")
-classpath = "polynote.jar:" + ":".join([":".join([ f'"{d}"' for d in deps ]), ":".join([ f'"{p}"' for p in plugins ])])
-cmd = f"java -cp polynote.jar:{classpath} -Djava.library.path={jep_path} polynote.Main {' '.join(sys.argv[1:])}"
```
Comment (jonathanindig) [inline scripts/polynote.py:34]: Is `polynote.jar` no longer needed because it ends up in `polynote_dir`?

File: polynote-kernel/src/main/scala/polynote/kernel/remote/transport.scala [hunk 6]
```diff
@@ -324,10 +327,45 @@ object SocketTransport {
       }
     }
 
+    private def findScalaVersion: URIO[BaseEnv with GlobalEnv with CurrentNotebook, String] = for {
+      serverConfig   <- Config.access
+      notebookConfig <- CurrentNotebook.config
+      scalaVersion   <- ZIO.succeed(notebookConfig.scalaVersion).some
+        .orElse(ZIO.succeed(serverConfig.kernel.scalaVersion).some)
+        .orElse(deployCommand.detectScalaVersion.some)
```
Comment (jonathanindig) [inline polynote-kernel/src/main/scala/polynote/kernel/remote/transport.scala:335]: What do you think about potentially adding the Scala version to `SparkPropertySet`? If Scala versions vary by property that seems like a logical place to specify that (and could solve possible issues if `spark-submit` version detection doesn't work for some reason.

Comment (jeremyrsmith) [reply polynote-kernel/src/main/scala/polynote/kernel/remote/transport.scala:335]: That's a good point! I agree it's a natural place to configure the Scala version.
